content/advent-2017/kubernetes-ready-service.md

```diff
@@ -0,0 +1,763 @@
+
+### Step 11. Dockerfile
+
+Our application is almost ready to be run in Kubernetes. Now we need to dockerize it.
+
+The simplest `Dockerfile` we need to define here might look like this:
+
+`Dockerfile`
+```docker
+FROM scratch
+
+ENV PORT 8000
+EXPOSE $PORT
+
+COPY advent /
+CMD ["/advent"]
+```
+
+We create the smallest container, copy the binary there and run it (we also do not forget about `PORT` configuration variable).
+
+Let's change a bit the `Makefile` to be able to build and image and run a container. Here it might be usefule to define new variables: `GOOS` and `GOARCH` which we will use for cross-compilation in the `build` goal. 
+
+`Makefile`
+```make
+...
+
+GOOS?=linux
+GOARCH?=amd64
+
+...
+
+build: clean
+	CGO_ENABLED=0 GOOS=${GOOS} GOARCH=${GOARCH} go build \
+		-ldflags "-s -w -X ${PROJECT}/version.Release=${RELEASE} \
+		-X ${PROJECT}/version.Commit=${COMMIT} -X ${PROJECT}/version.BuildTime=${BUILD_TIME}" \
+		-o ${APP}
+
+container: build
+	docker build -t $(APP):$(RELEASE) .
+
+run: container
+	docker stop $(APP):$(RELEASE) || true && docker rm $(APP):$(RELEASE) || true
+	docker run --name ${APP} -p ${PORT}:${PORT} --rm \
+		-e "PORT=${PORT}" \
+		$(APP):$(RELEASE)
+
+...
+```
+
+We also added the `container` goal to be able to build an image and the `run` goal to run our application from the container. All changes are available [here](https://github.com/rumyantseva/advent-2017/commit/909fef6d585c85c5e16b5b0e4fdbdf080893b679).
+
+Now let's try `make run` to check the whole process.
+
+### Step 12. Vendoring
+
+We have an external dependency (`github.com/gorilla/mux`) in our project. And it means that for production readiness we definetely need to [add dependency management here](https://github.com/rumyantseva/advent-2017/commit/7ffa56a78400367e5d633521dee816b767d7d05d). If we use [dep](https://github.com/golang/dep) the only thing which we need for our service is `dep init`:
+
+```bash
+$ dep init
+  Using ^1.6.0 as constraint for direct dep github.com/gorilla/mux
+  Locking in v1.6.0 (7f08801) for direct dep github.com/gorilla/mux
+  Locking in v1.1 (1ea2538) for transitive dep github.com/gorilla/context
+```
+
+It created `Gopkg.toml` and `Gopkg.lock` files and `vendor` directory. Personally, I prefer to push `vendor` to git, especially for important projects.
+
+### Step 13. Kubernetes
+
+[The last step](https://github.com/rumyantseva/advent-2017/commit/27b256191dc8d4530c895091c49b8a9293932e0f). Let's run our application in Kubernetes. The simplest way to run it locally is to install and configure [minikube](https://github.com/kubernetes/minikube).
+
+Kubernetes pull images from a Docker registry. In our case, we will work with the public Docker registry - [Dockerhub](https://hub.docker.com). We need to add to the `Makefile` one more variable and one more command:
```

Comment:
Suggestions: `pull` should be `pulls`.

Reply:
Also, it is more common in English to say "We need to add one more variable and one more command to the `Makefile`:", but your original sentence make perfect sense as it is.